compiler/luci/pass/src/FuseHorizontalFCLayers.cpp

```diff
@@ -0,0 +1,178 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FuseHorizontalFCLayers.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Service/Nodes/CircleConst.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+namespace
+{
+bool check_type_and_shape_equality(luci::CircleNode *left, luci::CircleNode *right)
+{
+  if (left->dtype() != right->dtype())
+    return false;
+
+  if (left->rank() != right->rank())
+    return false;
+
+  for (uint32_t i = 0; i < left->rank(); ++i)
+  {
+    if (left->dim(i).value() != right->dim(i).value())
+      return false;
+  }
+
+  return true;
+}
+
+template <loco::DataType D> void sum_const_values(luci::CircleConst *left, luci::CircleConst *right)
+{
+  const auto size = left->template size<D>();
+
+  for (uint32_t i = 0; i < size; ++i)
+  {
+    left->template at<D>(i) += right->template at<D>(i);
+  }
+}
+
+bool fuse_horizontal_fc_nodes(luci::CircleAdd *add_node)
+{
+  // Let's check left and right FC nodes
+  auto left_fc_node = dynamic_cast<luci::CircleFullyConnected *>(add_node->x());
+  auto right_fc_node = dynamic_cast<luci::CircleFullyConnected *>(add_node->y());
+
+  if (left_fc_node == nullptr or right_fc_node == nullptr)
+    return false;
+
+  if (not check_type_and_shape_equality(left_fc_node, right_fc_node))
+    return false;
+
+  if (left_fc_node->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  if (right_fc_node->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return false;
+
+  // Let's check that FC nodes have the same input
+  auto left_input_node = dynamic_cast<luci::CircleNode *>(left_fc_node->input());
+  if (left_input_node == nullptr)
+    return false;
+
+  auto right_input_node = dynamic_cast<luci::CircleNode *>(right_fc_node->input());
+  if (right_input_node == nullptr)
+    return false;
+
+  if (left_input_node != right_input_node)
+    return false;
+
+  // Lets check left and right FC weights: type and shape
+  auto left_fc_weights = dynamic_cast<luci::CircleConst *>(left_fc_node->weights());
+  auto right_fc_weights = dynamic_cast<luci::CircleConst *>(right_fc_node->weights());
+
+  if (left_fc_weights == nullptr or right_fc_weights == nullptr)
+    return false;
+
+  if (not check_type_and_shape_equality(left_fc_weights, right_fc_weights))
+    return false;
+
+  // Lets check left and right FC bias: type and shape
+  auto left_fc_bias = dynamic_cast<luci::CircleConst *>(left_fc_node->bias());
+  auto right_fc_bias = dynamic_cast<luci::CircleConst *>(right_fc_node->bias());
+
+  if ((left_fc_bias == nullptr and right_fc_bias != nullptr) or
+      (left_fc_bias != nullptr and right_fc_bias == nullptr))
+  {
+    return false;
+  }
+
+  if (left_fc_bias != nullptr and not check_type_and_shape_equality(left_fc_bias, right_fc_bias))
+    return false;
+
```

Comment:
Need to check if non-const bias is CircleOutputExclude.

```suggestion
  if (left_fc_bias == nullptr)
  {
    auto left_no_bias = dynamic_cast<luci::CircleOutputExclude *>(left_fc_node->bias());
    auto right_no_bias = dynamic_cast<luci::CircleOutputExclude *>(right_fc_node->bias());
    if (not left_no_bias or not right_no_bias)
      return false;
  }
```

Reply:
Done